File: java-boot/src/main/java/se/callistaenterprise/kotlindsl/repo/CustomerRepository.java
package se.callistaenterprise.kotlindsl.repo;

import org.jooq.DSLContext;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;
import se.callistaenterprise.kotlindsl.model.Customer;
import se.callistaenterprise.kotlindsl.model.CustomerIn;

import static org.jooq.Records.mapping;
import static se.callistaenterprise.kotlindsl.db.Tables.CUSTOMER;

@Repository
public class CustomerRepository {

    private final DSLContext jooq;

    public CustomerRepository(DSLContext jooq) {
        this.jooq = jooq;
    }

    public Customer getById(long id) {
        return jooq.select(CUSTOMER.ID, CUSTOMER.CUSTOMER_ID, CUSTOMER.STATUS, CUSTOMER.FIRST_NAME, CUSTOMER.LAST_NAME)
                .from(CUSTOMER)
                .where(CUSTOMER.ID.eq(id))
                .fetchOne(mapping(Customer::new));
    }

    @Transactional
    public long create(CustomerIn customer) {
        return jooq.insertInto(CUSTOMER, CUSTOMER.CUSTOMER_ID, CUSTOMER.STATUS, CUSTOMER.FIRST_NAME, CUSTOMER.LAST_NAME)
                .values(customer.customerId(), customer.status(), customer.firstName(), customer.lastName())
                .returningResult(CUSTOMER.ID).fetchOne().into(long.class);
    }

}

File: kotlin-boot/src/main/kotlin/se/callistaenterprise/kotlindsl/repo/CustomerRepository.kt
package se.callistaenterprise.kotlindsl.repo

import org.jooq.DSLContext
import org.jooq.Records.mapping
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional
import se.callistaenterprise.kotlindsl.db.Tables.CUSTOMER
import se.callistaenterprise.kotlindsl.model.Customer
import se.callistaenterprise.kotlindsl.model.CustomerIn
import se.callistaenterprise.kotlindsl.model.CustomerStatus

@Repository
class CustomerRepository(private val jooq: DSLContext) {

    fun getById(id: Long): Customer? {
        with(CUSTOMER) {
            return jooq.select(ID, CUSTOMER_ID, STATUS, FIRST_NAME, LAST_NAME)
                .from(CUSTOMER)
                .where(CUSTOMER.ID.eq(id))
                .fetchOne(mapping(::Customer))
        }
    }

    @Transactional
    fun create(customer: CustomerIn): Long {
        val id = jooq.newRecord(CUSTOMER, customer).apply {
            status = CustomerStatus.ACTIVE
            insert()
        }.id!!
        return id
    }

}
